README.md
# Wacktrace

Make your stacktraces wack! Insert arbitrary content into the call stack if that is, for some reason, something you want to do. Who am I to judge?

```
# in example.rb:
Wacktrace.add_to_stack([
  ['Fire in the disco!', 1, '🔥🕺 '],
  ['Fire in the Taco Bell!', 2, '🔥🌮 '],
  ['Fire in the disco!', 3, '🔥🕺 '],
  ['Fire in the gates of hell!', 4, '🔥😈 '],
]) do
  raise "High Voltage!"
end
```

```
$ ruby example.rb
example.rb:25:in `block in <main>': High Voltage! (RuntimeError)
	from /Users/kevin/.rvm/gems/ruby-3.0.0/gems/wacktrace-0.1.0/lib/wacktrace.rb:49:in `block in add_to_stack'
	from 🔥🕺 :1:in ` Fire in the disco︕ '
	from 🔥🌮 :2:in ` Fire in the Taco Bell︕'
	from 🔥🕺 :3:in ` Fire in the disco︕'
	from 🔥😈 :4:in ` Fire in the gates of hell︕'
	from /Users/kevin/.rvm/gems/ruby-3.0.0/gems/wacktrace-0.1.0/lib/wacktrace.rb:50:in `add_to_stack'
	from example.rb:19:in `<main>'
```

Amaze your friends! Annoy your coworkers! Ok, it's mostly that second one! But maybe someone has a good use for this, like inserting warnings into stack traces:

```
Traceback (most recent call last):
	14: from example.rb:47:in `<main>'
	13: from example.rb:35:in `dangerous_method'
	12: from /Users/kevin/.rvm/gems/ruby-2.7.2/gems/wacktrace-0.1.0/lib/wacktrace.rb:60:in `add_to_stack_from_lyrics'
	11: from /Users/kevin/.rvm/gems/ruby-2.7.2/gems/wacktrace-0.1.0/lib/wacktrace.rb:50:in `add_to_stack'
	10: from !:in ` '
	 9: from !:1:in `  ––––––––––––––––– WARNING︕ ––––––––––––––––––––––'
	 8: from !:2:in ` ｜ If you hit an error here， you ＊really＊ need ｜'
	 7: from !:3:in ` ｜ to go over to the FooBarBaz admin panel and    ｜'
	 6: from !:4:in ` ｜ clean up the Spleem that were erroniously      ｜'
	 5: from !:5:in ` ｜ created․  If you don’t do that， bad things    ｜'
	 4: from !:6:in ` ｜ will happen︕                                  ｜'
	 3: from !:7:in `  ––––––––––––––––––––––––––––––––––––––––––––––––––'
	 2: from /Users/kevin/.rvm/gems/ruby-2.7.2/gems/wacktrace-0.1.0/lib/wacktrace.rb:49:in `block in add_to_stack'
	 1: from example.rb:44:in `block in dangerous_method'
example.rb:30:in `real_dangerous_method': terrible error (RuntimeError)
```

### How's it work?

Sketchily! It's a set of wack stack hacks. The basic idea is that we just set up a series of methods like:

```
def a; b; end
def b; c; end
def c; actual_function; end
```

But to dynamically generate methods, you'd usually reach for `define_method`. Unfortunately, methods defined with `define_method` don't actually show up in call stacks because that'd be too easy. Instead we'll use `eval("def #{method_name};#{next_method_name};end")`. And, as a bonus, `eval` lets us pass in a filename and line number for stack trace purposes!

But to avoid having a hundred methods in the global namespace like `baby_shark_do_do_do_do_do_do_do`, we'll `Class.new`-up a temporary class and then `instance_eval` on that.

But now we're stuck with underscore-laden method names. We're not gonna stand for that kinda limitation. We're gonna reach for the sun and slap the face of god. We'll take arbitrary text and replace all the invalid characters like `"` and `@` with unicode near-equivalents like `＂` and `＠`. Because `def Cthulhu R'lyeh wgah'nagl fhtagn` isn't a valid ruby method, but the visually-identical `def Cthulhu R’lyeh wgah’nagl fhtagn` is!

Other fun tricks:

- Detect which direction ruby is printing stack traces today (it's newest-first on a gibbous or waxing moon, and newest-last the rest of the month)
- Deal with duplicate method names by padding them with non-breaking-spaces
- Prepend all methods with non-breaking spaces so ruby doesn't get confused by methods that start with capital letters

## Installation

Add this line to your application's Gemfile:

```ruby
gem 'wacktrace'
```

And then execute:

    $ bundle install

Or install it yourself as:

    $ gem install wacktrace

## Usage

To have full(ish) control over the stack trace contents, call `add_to_stack` and give it a block. That block will be run with the lines you provide in its call stack:

```
Wacktrace.add_to_stack([
  ['method name 1', 123, 'file name 1'],
  ['method name 2', 123, 'file name 2'],
  ['method name 3', 123, 'file name 3'],
]) do
  puts caller_locations
  # Prints out:
  #  file name 1:123:in ` method name 1'
  #  file name 2:456:in ` method name 2'
  #  file name 3:789:in ` method name 3'
end
```

If you want a quicker way to shove a bunch of text (like a warning, comment, or classic poem on the fleeting nature of mankind's greatest achievements), you can use `add_to_stack_from_lyrics` with a single newline-delimited string:

```
lyrics = "I met a traveller from an antique land,
Who said—“Two vast and trunkless legs of stone
Stand in the desert... Near them, on the sand,
Half sunk a shattered visage lies, whose frown,
And wrinkled lip, and sneer of cold command,
Tell that its sculptor well those passions read
Which yet survive, stamped on these lifeless things,
The hand that mocked them, and the heart that fed;
And on the pedestal!, these words appear:
My name is Ozymandias, King of Kings;
Look on my Works, ye Mighty, and despair!
Nothing beside remains. Round the decay
Of that colossal Wreck, boundless and bare
The lone and level sands stretch far away.”"
Wacktrace.add_to_stack_from_lyrics(lyrics, 'Percy Bysshe Shelley') { raise "Suck it, Horace Smith." }
```

...which will result in...

```
ruby example.rb
example.rb:17:in `block in <main>': Suck it, Horace Smith. (RuntimeError)
	from /Users/kevin/.rvm/gems/ruby-3.0.0/gems/wacktrace-0.1.0/lib/wacktrace.rb:56:in `block in add_to_stack'
	from Percy Bysshe Shelley:in ` I met a traveller from an antique land，'
	from Percy Bysshe Shelley:1:in ` Who said—“Two vast and trunkless legs of stone'
	from Percy Bysshe Shelley:2:in ` Stand in the desert․․․ Near them， on the sand，'
	from Percy Bysshe Shelley:3:in ` Half sunk a shattered visage lies， whose frown，'
	from Percy Bysshe Shelley:4:in ` And wrinkled lip， and sneer of cold command，'
	from Percy Bysshe Shelley:5:in ` Tell that its sculptor well those passions read'
	from Percy Bysshe Shelley:6:in ` Which yet survive， stamped on these lifeless things，'
	from Percy Bysshe Shelley:7:in ` The hand that mocked them， and the heart that fed;'
	from Percy Bysshe Shelley:8:in ` And on the pedestal︕， these words appear：'
	from Percy Bysshe Shelley:9:in ` My name is Ozymandias， King of Kings;'
	from Percy Bysshe Shelley:10:in ` Look on my Works， ye Mighty， and despair︕'
	from Percy Bysshe Shelley:11:in ` Nothing beside remains․ Round the decay'
	from Percy Bysshe Shelley:12:in ` Of that colossal Wreck， boundless and bare'
	from Percy Bysshe Shelley:13:in ` The lone and level sands stretch far away․”'
	from /Users/kevin/.rvm/gems/ruby-3.0.0/gems/wacktrace-0.1.0/lib/wacktrace.rb:57:in `add_to_stack'
	from /Users/kevin/.rvm/gems/ruby-3.0.0/gems/wacktrace-0.1.0/lib/wacktrace.rb:67:in `add_to_stack_from_lyrics'
	from example.rb:17:in `<main>'
```

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake spec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and the created tag, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/[USERNAME]/wacktrace. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [code of conduct](https://github.com/[USERNAME]/wacktrace/blob/main/CODE_OF_CONDUCT.md).

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Code of Conduct

Everyone interacting in the Wacktrace project's codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/[USERNAME]/wacktrace/blob/main/CODE_OF_CONDUCT.md).


CHANGELOG.md
## [1.0.1] - 2021-08-30

- Fixed backtrace order detection so inserted lines wouldn't be reversed in rubys earlier than 3.0.

## [1.0.0] - 2021-08-30

- Initial release


spec/spec_helper.rb
# frozen_string_literal: true

require "wacktrace"
require "pry-byebug"

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = ".rspec_status"

  # Disable RSpec exposing methods globally on `Module` and `main`
  config.disable_monkey_patching!

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
end


spec/wacktrace_spec.rb
# frozen_string_literal: true

RSpec.describe Wacktrace do
  describe 'add_to_stack' do
    let(:lines) {
      [
        ["somemethod1", 111, "somefile1"],
        ["somemethod2", 222, "somefile2"],
        ["somemethod3", 333, "somefile3"],
      ]
    }
    let(:real_return_value) { 3 }
    let(:real) { proc {
      @caller_locations = caller_locations.map(&:to_s)
      real_return_value
    } }
    subject { described_class.add_to_stack(lines, &real) }

    it "adds lines to the stack trace" do
      subject
      expect(@caller_locations[1, 3]).to eq [
        "somefile1:111:in ` somemethod1'",
        "somefile2:222:in ` somemethod2'",
        "somefile3:333:in ` somemethod3'",
      ]
    end

    it "still calls the underlying block" do
      expect(subject).to eq real_return_value
    end

    context 'with no lines' do
      let(:lines) { [] }
      it "still calls the underlying block" do
        expect(subject).to eq real_return_value
      end
    end
    context 'with only one line' do
      let(:lines) {
        [["somemethod1", 111, "somefile1"]]
      }
      it "adds line to the stack trace" do
        subject
        expect(@caller_locations[1, 1]).to eq [
          "somefile1:111:in ` somemethod1'",
        ]
      end
      it "still calls the underlying block" do
        expect(subject).to eq real_return_value
      end
    end

    context 'with duplicate method names' do
      let(:lines) {[
        ["somemethod1", 123, "somefile1"],
        ["somemethod2", 123, "somefile2"],
        ["somemethod1", 123, "somefile1"],
      ]}

      it "includes all lines in stack trace" do
        subject
        expect(@caller_locations[1, 3]).to eq [
          "somefile1:123:in ` somemethod1 '",
          "somefile2:123:in ` somemethod2'",
          "somefile1:123:in ` somemethod1'",
        ]
      end

      context 'with many duplicates' do
        let(:lines) {[
          ["somemethod1", 123, "somefile1"],
          ["somemethod1", 123, "somefile1"],
          ["somemethod1", 123, "somefile1"],
          ["somemethod1", 123, "somefile1"],
        ]}
        it "includes all lines with increasing non-breaking spaces" do
          subject
          expect(@caller_locations[1, 4]).to eq [
            "somefile1:123:in ` somemethod1   '",
            "somefile1:123:in ` somemethod1  '",
            "somefile1:123:in ` somemethod1 '",
            "somefile1:123:in ` somemethod1'",
          ]
        end
      end
    end
  end

  describe 'add_to_stack_from_lyrics' do
    let(:real_return_value) { 3 }
    let(:real) { proc {
      @caller_locations = caller_locations.map(&:to_s)
      real_return_value
    } }
    let(:lyrics) { "
row row row your boat
gently down the stream
    ".strip }
    let(:filename) { "whatever" }
    subject { described_class.add_to_stack_from_lyrics(lyrics, filename, &real) }
    it "constructs lines and passes them to add_to_stack" do
      expect(described_class).to receive(:add_to_stack).with(
        [
          ["row row row your boat", 0, filename],
          ["gently down the stream", 1, filename]
        ]
      )
      subject
    end
  end

  describe 'detect_order' do
    it 'detects order' do
      # Passes in ruby 3.0 only
      expect(Wacktrace.send(:detect_order)).to eq :recent_first
    end

    it 'detects recent first order' do
      allow_any_instance_of(Exception).to receive(:full_message) { "detect_order_raise\ndetect_order\nother_method"}
      expect(Wacktrace.send(:detect_order)).to eq :recent_first
    end
    it 'detects recent last order' do
      allow_any_instance_of(Exception).to receive(:full_message) { "other_method\ndetect_order\ndetect_order_raise"}
      expect(Wacktrace.send(:detect_order)).to eq :recent_last
    end
  end
end


lib/wacktrace.rb
# frozen_string_literal: true

require_relative "wacktrace/version"

module Wacktrace
  class Error < StandardError; end
  class << self
    # Runs the given block with the given lines inserted into the call stack
    # above it.  Lines should be an array like:
    #
    # [
    #   ["somemethod1", 111, "somefile1"],
    #   ["somemethod2", 222, "somefile2"],
    #   ["somemethod3", 333, "somefile3"],
    # ]
    #
    # That will result in a strack trace like:
    #
    #  "somefile3:333:in ` somemethod3'",
    #  "somefile2:222:in ` somemethod2'",
    #  "somefile1:111:in ` somemethod1'",
    #
    def add_to_stack(lines, &real)
      if lines.length <= 0
        return real.call
      end

      # If stack traces are being printed newest-to-oldest, we need to reverse
      # the input lines in order to have them print out readably.
      if detect_order == :recent_first
        lines.reverse!
      end

      lines = fix_duplicate_methods(lines)

      # Create a namespace for all these methods so we don't polute the global
      # namespace.
      container_class = Class.new
    
      raise "add to stack missing block" unless block_given?
    
      lines = lines.map { |line| [clean_method_name(line[0]), line[1], line[2]] }
    
      # Define each method in series: a calls b, b calls c, and so on.  The last
      # line does not get a method defined here.
      lines.each_cons(2) do |line_1, line_2|
        # puts "defining '#{line_1[0]}'"
        success = true
        define_stack_level(container_class, line_1, line_2[0])
      end
    
      last = lines.last
      define_stack_level(container_class, last, 'ending')
      container_class.define_singleton_method("ending") { real.call }
      return container_class.send(lines.first[0])
    end

    # This is a convenience method to construct a stack trace from a single
    # string with a bunch of newlines (eg the lyrics to a song or words to a
    # poem).  It'll use the same filename for each line.
    def add_to_stack_from_lyrics(lyrics, filename, &block)
      lines = lyrics.split("\n").map.with_index do |line, i|
        [line, i, filename]
      end
      add_to_stack(lines, &block)
    end

    private

    # Because we can't have two methods with the same name in the same
    # namespace, we'll need to modify any lines with the same method name
    # (line[0]).  To do this, we'll just pad each duplicate line with
    # non-breaking spaces.  So if you have several lines with the method name
    # "foo", they'll become "foo", "foo ", "foo  ", etc.
    def fix_duplicate_methods(lines)
      counts = {}
      lines.map { |line|
        method_name = line[0]
        if counts[method_name]
          original_method_name = method_name
          method_name = method_name + (' ' * counts[method_name])
          counts[original_method_name] += 1
        else
          counts[method_name] = 1
        end
        [method_name, line[1], line[2]]
      }
    end

    def clean_method_name(string)
      # Always start the method name with a non-breaking space so ruby doesn't
      # interpret it as a constant if you started the string with a capital.
      ' ' + string
        .gsub(' ', ' ')
        .gsub(',', '，')
        .gsub('.', '․')
        .gsub(':', '：')
        .gsub("'", '’')
        .gsub("@", '＠')
        .gsub("`", '`')
        .gsub("~", '～')
        .gsub("#", '＃')
        .gsub("$", '＄')
        .gsub("%", '％')
        .gsub("^", '＾')
        .gsub("&", '&')
        .gsub("*", '＊')
        .gsub("(", '（')
        .gsub(")", '）')
        .gsub("-", '–')
        .gsub("+", '＋')
        .gsub("[", '［')
        .gsub("]", '］')
        .gsub("|", '｜')
        .gsub("/", '／')
        .gsub('\\', '＼')
        .gsub(";", ';')
        .gsub("{", '｛')
        .gsub("}", '｝')
        .gsub("\"", '＂')
        .gsub("'", '＇')
        .gsub("<", '＜')
        .gsub(">", '＞')
        # These are technically allowed at the end of a method, but I'm too lazy to
        # write logic to allow that.
        .gsub('!', '︕')
        .gsub('?', '︖')
        .gsub('=', '＝')
    end

    # Calls the given block with the given "lines" inserted into the call stack.
    # Each line should look like:
    #
    #  [ "somemethodname", 123, "somefilename"]
    #
    # which will result in a traceback line that looks like:
    #
    #  3: from somemethodname:123:in `somefilename'
    #
    def define_stack_level(container_class, line, body)
      success = true

      # For some reason line numbers come out as one higher than whatever we put in here.
      line_number = line[1].to_i - 1
      method_name = line[0]
      file_name = line[2]

      method_string = "def #{method_name}\n#{body}\nend"

      begin
        container_class.instance_eval(method_string, file_name, line_number)
      rescue SyntaxError => e
        success = false
      end

      if !success || !container_class.methods.include?(method_name.to_sym)
        warn("
          Unable to create a method named:
          #{method_name}
          with body
          #{method_string}
          One of them probably contains some special character that's not yet handled."
        )
      end
    end

    def add_lines_to_stack(lines, &real)
      add_to_stack(lyrics_to_stack_lines(lines), &real)
    end

    # The backtrace order varies by situation (version, tty vs file, among
    # others). Rather than hardcoding the rules, let's just detect it.
    def detect_order
      begin
        detect_order_raise
      rescue Exception => e
        trace_lines = e.full_message.split("\n")
        if trace_lines.first.include?("detect_order_raise")
          return :recent_first
        elsif trace_lines.last.include?("detect_order_raise")
          return :recent_last
        else
          return :undetectable
        end
      end
    end
    def detect_order_raise
      raise 'error'
    end
  end
end